GitUI/Editor/FileViewer.cs

```diff
@@ -1847,7 +1818,7 @@ private void PreviousChangeButtonClick(object sender, EventArgs e)
             {
                 string lineContent = internalFileViewer.GetLineText(line);
 
-                if (lineContent.StartsWithAny(inChange))
+                if (internalFileViewer.IsSearchMatch(lineContent))
```

Comment:
For symmetry reasons, I would expect:
```suggestion
                if ((_viewMode == ViewMode.RangeDiff) ^ internalFileViewer.IsSearchMatch(lineContent))
```

Reply:
For range-diff,bthis is the check to get to next header.
For normal diff, this is to stay in diff, to find when to look for the next section.

One reason why I split this in next pr.

Reply:
> One reason why I split this in next pr.

👍

---

I am not convinced that the first commit is a correct refactoring. It should be fixed since you want to keep the single commits.
`IsSearchMatch` is **!**`line.StartsWithAny(inChange)` for range diff.

Reply:
private static string[] _rangeDiffPrefixes = { "    " };

has opposite meaning to

        private static string[] _combinedDiffPrefixes = { "+", "-", " +", " -" };
        private static string[] _normalDiffPrefixes = { "+", "-" };

After the update

rangediff
public override bool IsSearchMatch(string line)
        {
            return !line.StartsWithAny(_diffSearchPrefixes);
        }

other

public virtual bool IsSearchMatch(string line)
        {
            return line.StartsWithAny(_diffSearchPrefixes);
        }

Same code in the second commit.

--

The second problem is what I described, that the code is doing a different thing, changed in the second commit.

Reply:
Since you do not want to squash the commits, i.e. to keep the first refactoring commit as a standalone commit, this commit shall not change the behavior - regardless of what is improved in the second commit.
Old code:
`                if (lineContent.StartsWithAny(inChange))`
The new code
`                if (internalFileViewer.IsSearchMatch(lineContent))`
ignores that `IsSearchMatch` is inverse for range diff.
Hence I expect it to be:
`                if ((_viewMode == ViewMode.RangeDiff) ^ internalFileViewer.IsSearchMatch(lineContent))`

Reply:
The invert is in `IsSearchMatch()`

https://github.com/gerhardol/gitextensions/blob/feature/i4912-git-grep-prepare/GitUI/Editor/Diff/DiffHighlightService.cs#L33
https://github.com/gerhardol/gitextensions/blob/feature/i4912-git-grep-prepare/GitUI/Editor/Diff/RangeDiffHighlightService.cs#L32

Reply:
I tested this again, range-diff work next/previous moves to the headers (as before, instead of the last).

Reply:
> I tested this again, range-diff work next/previous moves to the headers (as before, instead of the last).

Have you tested the first commit only?

> The invert is in IsSearchMatch()

I had noticed it before, and did not come to a conclusion yesterday. But today, I can point to it.
I still do not get why line 1821 is not changed the same way as line 1804:

![image](https://github.com/gitextensions/gitextensions/assets/36601201/3d11c744-9a3c-402f-a8c1-5631b4d61252)